File: packages/worker-contracts/src/worker_contracts/github.py
"""Versionierte Boundary-DTOs für github-service (ADR-0004 §1).

Was hier fehlt, ist die Aussage: **keine Punktzahl, kein Rang, keine
abgeleitete Eigenschaft** (ADR-0022). Ein Repository ist ein Beleg mit einem
Link, und die Bewertung überlässt dieses System dem Menschen, der ihn anklickt.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

__all__ = ["ConnectGitHubV1", "GitHubConnectionV1", "RepositoryV1"]


class ConnectGitHubV1(BaseModel):
    login: str = Field(..., min_length=1, max_length=39)


class RepositoryV1(BaseModel):
    name: str
    description: str
    #: Was GitHub als Hauptsprache meldet — weitergegeben, nicht ausgewertet.
    language: str | None
    stars: int
    url: str
    pushed_at: datetime | None


class GitHubConnectionV1(BaseModel):
    subject_id: UUID
    login: str
    verified: bool
    #: Steht nur in der eigenen Ansicht: die Einmalzeichenfolge nützt nur der
    #: Person, die den Gist anlegt.
    challenge_description: str | None = None
    #: Wann der Abzug geholt wurde. Die Anzeige sagt „Stand: …" — es wird nicht
    #: zugesehen, sondern einmal gelesen.
    fetched_at: datetime | None = None
    repositories: list[RepositoryV1] = Field(default_factory=list)

File: web/src/features/company/api/github.ts
// Die GitHub-Belege einer Person, wie ein Unternehmen sie sieht.
//
// Was hier fehlt, ist die Aussage: keine Punktzahl, kein Rang, keine
// abgeleitete Eigenschaft (ADR-0022). Ein Repository ist ein Beleg mit einem
// Link — nachprüfbar, ohne Zwischenrechnung, und widersprechbar.
//
// ACHTUNG, gemeldeter Widerspruch: `GITHUB_BASE_URL` fällt in `src/env.ts` auf
// Port **8010** zurück; github-service läuft auf **8011** (8010 ist
// notification-service). `env.ts` liegt außerhalb dieses Territoriums und wurde
// nicht angefasst — steht in `docs/uebergabe/frontend-work-company.md`.

import { request } from "../../../core/api/client";
import { GITHUB_BASE_URL } from "../../../env";
import { type Fehlschlag, deuten } from "../../../shared/api/fehler";

export interface GitHubRepository {
  name: string;
  description: string;
  language: string | null;
  stars: number;
  url: string;
  pushed_at: string | null;
  /** Die MENGE der Sprachen, nie ihr Anteil (ADR-0022 §2). */
  languages: string[];
  /** Die Topics, die der Besitzer selbst gesetzt hat — eine Nennung. */
  topics: string[];
}

export interface GitHubConnection {
  subject_id: string;
  login: string;
  verified: boolean;
  fetched_at: string | null;
  repositories: GitHubRepository[];
}

/**
 * Die Verbindung einer anderen Person.
 *
 * `404` wird zu `null` — „gibt es nicht", „nicht bewiesen" und „nicht
 * freigegeben" sind für die Oberfläche derselbe Fall, genau wie der Server sie
 * ununterscheidbar hält.
 */
export async function getGitHub(
  subjectId: string,
  signal?: AbortSignal
): Promise<{ ok: true; connection: GitHubConnection | null } | Fehlschlag<"unavailable">> {
  const answer = await request<GitHubConnection>(
    GITHUB_BASE_URL,
    `/github/${subjectId}`,
    { signal },
    "fehler.verbindungNichtGeladen"
  );
  if (answer.ok) return { ok: true, connection: answer.value ?? null };
  if (answer.error.status === 404) return { ok: true, connection: null };
  return deuten<"unavailable">(
    answer.error,
    { 503: { reason: "unavailable", titel: "fehler.ledgerSchweigt" } },
    "unavailable"
  );
}
